static/js/code-highlighter.js

```diff
@@ -8,6 +8,198 @@
+    this.currentTitle = this.originalTitle;
+
+    this.stickyTitle = null;
+    this.selectionTitle = null;
+  }
+
+  updateTitle() {
+    let bestTitle;
+
+    // If we have a better title than the original title, use it, but making
+    // sure to include the original filename because this is important for
+    // finding the tab again in the awesomebar via its filename.
+    if (this.selectionTitle) {
+      bestTitle = `${this.selectionTitle} (${this.originalTitle})`;
+    } else if (this.stickyTitle) {
+      bestTitle = `${this.stickyTitle} (${this.originalTitle})`;
+    } else {
+      bestTitle = this.originalTitle;
+    }
+
+    // Debounce setting the title out of a fear of slowing down scrolling but
+    // not wanting to to do the legwork to figure out if this would pose a
+    // problem.  I am assuming this shouldn't result in synchronous reflows.
+    if (bestTitle === this.currentTitle) {
+      return;
+    }
+    document.title = this.currentTitle = bestTitle;
+  }
+
+  /**
+   * Recognizing that namespaces can be a little verbose and use up precious
+   * space, flatten things that look like namespaces to be a single-character
+   * delimited by a single colon.
+   *
+   *
+   * Good example transforms:
+   * - `Foo` => `Foo`
+   * - `Foo::Bar` => `Foo::Bar`
+   * - `mozilla::Foo::Bar` => `m:Foo::Bar`
+   * - `mozilla::dom::quota::Foo::Bar` => `m:d:q:Foo::Bar`
+   *
+   * Sketchy example transforms:
+   * - `mozilla::Foo` => `m:Foo`
+   * - `mozilla::dom::Foo` => `m:d:Foo`
+   *
+   * The sketchy transforms are sketchy because we're assuming that lowercase is
+   * indicative of a namespace.  Note that, similarly to the comments in
+   * `_findBestPrettySymbolInSourceLineElem`, this heuristic should simply be
+   * mooted by having the symbol dictionary provide a pre-computed concise
+   * pretty name for a symbol.  That mechanism can leverage knowing what is and
+   * isn't a namespace/class and also having global knowledge of the names that
+   * are in the codebase so that additional tokens can be used in cases where
+   * ambiguity exists, etc.
+   */
+  _shortenNamespaces(pretty) {
+    const pieces = pretty.split("::");
+    // Nothing to do if we don't have anything to split.
+    if (pieces.length < 2) {
+      return pretty;
+    }
+
+    // We want to use the last two components if they are both initial-caps,
+    // but not for the "mozilla::Foo" case, or "mozilla::dom::Foo" case, where
+    // we still want to collapse the lower-cased namespace.  In that case, we
+    // collapse everything but the last piece.
+    let splitPoint;
+    if (/^[a-z]/.test(pieces[pieces.length - 2])) {
+      splitPoint = -1;
+    } else {
+      splitPoint = -2;
+    }
+
+    const nsPieces = pieces.slice(0, splitPoint);
+    const fullPieces = pieces.slice(splitPoint);
+
+    const nsTransformed = nsPieces.map((piece) => {
+      return piece[0];
+    });
+
+    if (nsTransformed.length === 0) {
```

Comment:
I'd write this as `if (!nsTransformed.length)`, but no big deal either way.

Reply:
> I'd write this as `if (!nsTransformed.length)`, but no big deal either way.

This is a stylistic thing I've been doing (after reading a thing I no longer remember where I read it) where the explicit zero check feels more explicit/literate/in-domain even if it ends up equivalent to the negation.  I will sometimes just do the negation when it's more of a falsey check for whether optional arguments were provided, etc.

That said, I'm definitely amenable to conforming to prettier/eslint rules that cover this, especially if we check in a vscode config.  I will leave this as-is for now.